File: 25-State/src/go/no_quarter_state.go
package main

import "fmt"

type NoQuarterState struct {
	machine *GumballMachine
}

func (s *NoQuarterState) InsertQuarter() {
	fmt.Println("  You inserted a quarter")
	s.machine.setState(s.machine.hasQuarter)
}
func (s *NoQuarterState) EjectQuarter() {
	fmt.Println("  You haven't inserted a quarter")
}
func (s *NoQuarterState) TurnCrank() {
	fmt.Println("  You turned, but there's no quarter")
}
func (s *NoQuarterState) Dispense() {
	fmt.Println("  You need to pay first")
}
func (s *NoQuarterState) Name() string { return "NoQuarter" }

File: 25-State/src/rust/no_quarter_state.rs
use crate::state::State;
use crate::gumball_machine::GumballMachine;
use crate::has_quarter_state::HasQuarterState;

pub struct NoQuarterState;

impl State for NoQuarterState {
    fn insert_quarter(self: Box<Self>, _machine: &mut GumballMachine) -> Box<dyn State> {
        println!("  You inserted a quarter");
        Box::new(HasQuarterState) // transition
    }
    fn eject_quarter(self: Box<Self>, _machine: &mut GumballMachine) -> Box<dyn State> {
        println!("  You haven't inserted a quarter");
        self
    }
    fn turn_crank(self: Box<Self>, _machine: &mut GumballMachine) -> Box<dyn State> {
        println!("  You turned, but there's no quarter");
        self
    }
    fn dispense(self: Box<Self>, _machine: &mut GumballMachine) -> Box<dyn State> {
        println!("  You need to pay first");
        self
    }
    fn name(&self) -> &str {
        "NoQuarter"
    }
}
